### DISCUSSION TITLE
VAT Invoicing Feed

### DISCUSSION THREAD

**Discussion by KerryXEX** [abuzuhri/Amazon-SP-API-CSharp#72] [General]
Hi,

Is there a way to upload VAT Invoice files (pdf) via the "VAT Invoicing Feed
Enumeration value: _UPLOAD_VAT_INVOICE_" ?

I couldn't find this in the feeds api?

Thanks
Kerry.

**Comment by abuzuhri** [comment]
Hello

Right now our api not support generate xml for this feed but you can generate xml out side and submit it with our api and get results 


Also you can folk the project and code for generate xml and I will merge

**Reply by KerryXEX** [reply]
Hi,

thanks for your fast reply.
Would you have an example of how to submit custom XML content?
e.g. how do you submit the feedtype just as string?
`amazonConnection.Feed.SubmitFeed(xml, ??);`

Thanks
Kerry.

**Reply by abuzuhri** [reply]
Hello @KerryXEX 

If you need to submit feed the feed should be one of the feed in the list 
https://github.com/amzn/selling-partner-api-docs/blob/main/references/feeds-api/feedtype-values.md

```
        public enum FeedType
        {
            JSON_LISTINGS_FEED,
            POST_PRODUCT_DATA,
            POST_INVENTORY_AVAILABILITY_DATA,
            POST_PRODUCT_OVERRIDES_DATA,
            POST_PRODUCT_PRICING_DATA,
            POST_PRODUCT_IMAGE_DATA,
            POST_PRODUCT_RELATIONSHIP_DATA,
            POST_FLAT_FILE_INVLOADER_DATA,
            POST_FLAT_FILE_LISTINGS_DATA,
            POST_FLAT_FILE_BOOKLOADER_DATA_,
            POST_FLAT_FILE_CONVERGENCE_LISTINGS_DATA,
            POST_FLAT_FILE_PRICEANDQUANTITYONLY_UPDATE_DATA,
            POST_UIEE_BOOKLOADER_DATA,
            POST_STD_ACES_DATA,
            POST_ORDER_ACKNOWLEDGEMENT_DATA,
            POST_PAYMENT_ADJUSTMENT_DATA,
            POST_ORDER_FULFILLMENT_DATA,
            POST_INVOICE_CONFIRMATION_DATA,
            POST_EXPECTED_SHIP_DATE_SOD,
            POST_FLAT_FILE_ORDER_ACKNOWLEDGEMENT_DATA,
            POST_FLAT_FILE_PAYMENT_ADJUSTMENT_DATA,
            POST_FLAT_FILE_FULFILLMENT_DATA,
            POST_EXPECTED_SHIP_DATE_SOD_FLAT_FILE,
            POST_FULFILLMENT_ORDER_REQUEST_DATA,
            POST_FULFILLMENT_ORDER_CANCELLATION_REQUEST_DATA,
            POST_FBA_INBOUND_CARTON_CONTENTS,
            POST_FLAT_FILE_FULFILLMENT_ORDER_REQUEST_DATA,
            POST_FLAT_FILE_FULFILLMENT_ORDER_CANCELLATION_REQUEST_DATA,
            POST_FLAT_FILE_FBA_CREATE_INBOUND_PLAN,
            POST_FLAT_FILE_FBA_UPDATE_INBOUND_PLAN,
            POST_FLAT_FILE_FBA_CREATE_REMOVAL,
            RFQ_UPLOAD_FEED,
            POST_EASYSHIP_DOCUMENTS
        }
```

**Reply by abuzuhri** [reply]
After i do searching i found this answer 

https://github.com/amzn/selling-partner-api-docs/issues/532#issuecomment-845273585


`"Please refer to the official MWS and SP-API documentation to check the up-to-date information, VAT invoice Uploading is not a valid feedtype, currently neither MWS API nor SP API support VAT invoice uploading:"`

**Reply by KerryXEX** [reply]
Hi,

I found the feed type in the official developer documentation here:
https://docs.developer.amazonservices.com/en_UK/feeds/Feeds_FeedType.html

And a hint that this was recently implemented by Amazon in this thread:
https://github.com/amzn/selling-partner-api-docs/issues/404

So it should be working by first creating the feed document, then upload the pdf to the document address and then submit the feed item with the corresponding type and document link.

**Reply by abuzuhri** [reply]
> I found the feed type in the official developer documentation here:
> https://docs.developer.amazonservices.com/en_UK/feeds/Feeds_FeedType.html

This  documentation for MWS not for SP-API

> And a hint that this was recently implemented by Amazon in this thread:
> amzn/selling-partner-api-docs#404

I will send spend some times in the weekend to implement this and try if its will work

**Reply by KerryXEX** [reply]
Hi Tareq,
Thanks alot, that would be great. I'm happy to support with testing.

As Amazon recently posted that they will measure seller quality also based on availability of VAT invoices, this will be helpful for many users I suppose.

**Reply by abuzuhri** [reply]
I try to submit feed with same way i submit others feed but i get this error

`{
  "errors": [
    {
      "code": "Unauthorized",
      "message": "Access to the resource is forbidden",
      "details": ""
    }
  ]
}`



Please download project NOT NUGET

and try this code from your side
```Csharp
            AmazonConnection amazonConnection = new AmazonConnection(new AmazonCredential()
            {
                AccessKey = Environment.GetEnvironmentVariable("AccessKey"),
                SecretKey = Environment.GetEnvironmentVariable("SecretKey"),
                RoleArn = Environment.GetEnvironmentVariable("RoleArn"),
                ClientId = Environment.GetEnvironmentVariable("ClientId"),
                ClientSecret = Environment.GetEnvironmentVariable("ClientSecret"),
                RefreshToken = Environment.GetEnvironmentVariable("RefreshToken"),
                MarketPlace = MarketPlace.UnitedArabEmirates, //MarketPlace.GetMarketPlaceByID("A2VIGQ35RCS4UG")
                IsActiveLimitRate = true
            });

            var options = new AmazonSpApiSDK.Models.Feeds.FeedOptions();
            options.Add("DocumentType", "Invoice");
            options.Add("InvoiceNumber", "FV-EUR/1/2/3");
            options.Add("OrderId", "???-???????-?????");
            options.Add("TotalAmount", "39.37");
            options.Add("TotalVATAmount", "6.29");


            var feedresultPDF = amazonConnection.Feed.SubmitFeed(@"C:\Users\tareq\Downloads\224129324d8e2096ec0a70f223572eda36.pdf"
                                                    , FeedType.UPLOAD_VAT_INVOICE
                                                    , new List<string>() { MarketPlace.UnitedArabEmirates.ID }
                                                    , options
                                                    , ContentType.PDF);
```

**Reply by KerryXEX** [reply]
Hi Tareq,

thanks alot - will try that now

**Reply by KerryXEX** [reply]
Yes, I get the same "unauthorized/forbidden" error.
While walkting through the steps it looks like the file upload works and the actual error occurs when submitting the XML feed request (at #106 in RequestService.cs).

I checked again the above link to the successful implementation and corresponding Amazon instructions (https://m.media-amazon.com/images/G/01/B2B/DeveloperGuide/vat_calculation_service__dev_guide_H383rf73k4hsu1TYRH139kk134yzs.pdf) where it says you have to enrol to VCS Services (which we have), but then I also saw that in our Developer Profile the section "Tax Services" is not checked which also lists "creating tax invoices", so I assume this is the issue with the "not allowed" response.

I'll try to get this addtional role approved and can check again then.

Thanks
Kerry.

**Reply by abuzuhri** [reply]
Ok Good I add the feature in nuget v1.0.38 , please let me know if you need more help

**Reply by KerryXEX** [reply]
Thanks - I'll keep you updated!

**Reply by RichDrag** [reply]
Hi,
i tried your above sample today to upload an invoice. I also get the message "Access to the resource is forbidden"
Is this caused by the "Restricted roles"? I currently can't access the PII Data, but i create a ticket. But do i really have to fulfill all the requierments regarding data security, if i process just my own data? Currently we got all this PII Data by MWS over AMTU and nobody ask about our data security...
We also don't use the Vat calculation services from amazon,....

**Comment by zinsmeik** [comment]
Hi, 
I tried to upload a PDF Invoice with the SubmitFeed.

```
        private void UploadInvoicePDF(AmazonOrderInvoice uploadInvoice)
        {
            var options = new FikaAmazonAPI.AmazonSpApiSDK.Models.Feeds.FeedOptions();

            options.Add("DocumentType", "Invoice");
            options.Add("InvoiceNumber", uploadInvoice.InvoiceNumber);
            options.Add("OrderId", uploadInvoice.AmazonOrderID);

            var feedresultPDF = AMAConnection.Connection.Feed
                .SubmitFeed(
                    uploadInvoice.InvoiceFileName,
                    FikaAmazonAPI.Utils.Constants.FeedType.UPLOAD_VAT_INVOICE,
                    new List<string>() { FikaAmazonAPI.Utils.MarketPlace.Germany.ID },
                    options,
                    FikaAmazonAPI.Utils.Constants.ContentType.PDF);

            if(feedresultPDF.Length > 0)
            {
                
                Thread.Sleep(1000 * 30);

                var feedOutput = AMAConnection.Connection.Feed.GetFeed(feedresultPDF);
                var outPut = AMAConnection.Connection.Feed.GetFeedDocument(feedOutput.ResultFeedDocumentId);

                var reportOutpit = outPut.Url;
                
                RenameUploadedFile(uploadInvoice.InvoiceFileName);
            }
        }


```
When I check the URL of the Report Output ist shows an error.

```
Feed Processing Summary:
	Number of records processed		1
	Number of records successful		0

original-record-number	sku	error-code	error-type	error-message
1		79513	Error	Please provide shipmentId or orderid
1		79513	Error	Please provide taxDocumentNumber
1		79532	Error	The invoice number you have passed contains special characters that are not supported by Amazon

```
It seams the option are not correctly sumitted to the feed.

Any hint on this?

Thanks
